Fixes #2721 - Allow use of spice for libvirt via existing Setting

GregSutcliffe opened the pull request:
I appreciate this is a slight regression as the Spice variant of `display` has `:listen => '0'` in it, but I wanted to open discussions before going to the effort of adding a whole new Setting. Opinions please?

GregSutcliffe committed:
Fixes #2721 - Allow use of spice for libvirt via checkbox per-CR
Fixes #2742 - Add a per-CR checkbox for VNC passwords

domcleal commented on lib/foreman/model/libvirt.rb:
```diff
@@ -134,12 +136,18 @@ def disconnect
 
     def vm_instance_defaults
+      display = if Setting[:libvirt_default_console_address] == 'spice'
+                  { :type => 'spice', :listen => '0', :port => '-1' }
```
Shouldn't we set a password too?

ohadlevy replied:
and name..

domcleal commented:
Perhaps add this as an attribute to the CR rather than a new global setting?

GregSutcliffe commented:
That's a good idea. I'll see if I can arrange that next week. Also solves the re-use of `:display` which I wasn't actually that happy with.

GregSutcliffe commented:
Updated to be per-CR and added a new feature to control whether we set a display password per-CR. Seems to work here (although setting passwords on Spice displays results in a Permission Denied error for me).

There's probably a better way to handle the optional display components on show.html.erb, ideas?

ohadlevy commented on lib/foreman/model/libvirt.rb:
```diff
@@ -3,6 +3,22 @@ class Libvirt < ComputeResource
+    end
+
+    def set_password=(setpw)
+      self.attrs = self.attrs.merge({:setpw=>setpw})
```
why not

``` ruby
self.attrs[:setpw] = setpw
```

GregSutcliffe replied:
Just me cut'n'pasting ideas from the NIC models, which use merge. I can change it if people prefer, I don't have a strong preference.

domcleal commented on app/views/compute_resources/show.html.erb:
```diff
@@ -32,6 +32,16 @@
         <td><%= _("URL") %></td>
         <td><%= @compute_resource.url %></td>
       </tr>
+      <% if @compute_resource.type == "Foreman::Model::Libvirt" -%>
```
Eugh :)

How about copying the way form works and have smaller templates we can render here?

GregSutcliffe replied:
Eugh indeed. Is it worth breaking it down into even smaller forms? Wouldn't we still need some form of conditional on rendering?

domcleal replied:
The others use:

```
render "compute_resources/form/#{@compute_resource.provider.downcase}"
```

so no conditional.  I think it's worth doing if this is the alternative.  We will probably start finding a use for it in other compute resources if the infrastructure's there.

domcleal commented on lib/foreman/model/libvirt.rb:
```diff
@@ -134,12 +153,19 @@ def disconnect
+        :display    => {
+                         :type => display_type,
+                         :listen => Setting[:libvirt_default_console_address],
+                         :password => random_password,
```
Should `set_password?` be checked here?  What happens if we pass `:password => nil` in, does it ignore it or set no password?  (I think we would want the former, hence I wonder if we should move the check.)

GregSutcliffe replied:
From my tests, we get the correct behaviour - if set_password is false, a password is not set. I don't see why we'd end up with :password => nil if set_password is true - the random_password function should handle that.

domcleal replied:
I was thinking if set_password was false, whether `:password => nil` would cause an empty password to be set.  If it doesn't for VNC and SPICE, then that's fine, thanks.

domcleal commented on app/views/compute_resources/form/_libvirt.html.erb:
```diff
@@ -1,5 +1,8 @@
 <%= text_f f, :url, :class => "input-xlarge", :help_block => _("e.g. qemu://host.example.com/system") %>
 
+<%= checkbox_f f, :use_spice, :checked => f.object.use_spice?, :help_inline => _("Use the Spice display protocol on this hypervisor?") %>
```
SPICE, not Spice.

I think a dropdown of display types would be better, so it's clearer that VNC is the other option when this is unchecked.

GregSutcliffe replied:
Yeah, makes sense

domcleal commented on app/views/compute_resources/form/_libvirt.html.erb:
```diff
@@ -1,5 +1,8 @@
 <%= text_f f, :url, :class => "input-xlarge", :help_block => _("e.g. qemu://host.example.com/system") %>
 
+<%= checkbox_f f, :use_spice, :checked => f.object.use_spice?, :help_inline => _("Use the Spice display protocol on this hypervisor?") %>
+<%= checkbox_f f, :set_password, :checked => f.object.set_password?, :help_inline => _("Set a randomly generated password on the VNC display") %>
```
It's not VNC if it's SPICE.

domcleal commented on lib/foreman/model/libvirt.rb:
```diff
@@ -100,9 +117,11 @@ def console uuid
-      # Listen address cannot be updated while the guest is running
-      # When we update the display password, we pass the existing listen address
-      vm.update_display(:password => password, :listen => vm.display[:listen])
+      unless vm.display[:type] =~ /spice/i
```
Why?

GregSutcliffe replied:
This _may_ be a leftover from when I wasn't setting a SPICE password. I'll retest without the conditional.

GregSutcliffe commented:
@domcleal updated to address your points. I _think_ it works, but I've some issues with my SPICE setup atm - could you at least check I haven't broken the VNC parts?

GregSutcliffe commented on app/models/compute_resources/foreman/model/libvirt.rb:
```diff
@@ -102,7 +119,7 @@ def console uuid
       # Listen address cannot be updated while the guest is running
       # When we update the display password, we pass the existing listen address
-      vm.update_display(:password => password, :listen => vm.display[:listen])
+      vm.update_display(:password => password, :listen => vm.display[:listen]) unless vm.display[:type] =~ /spice/i
```
@domcleal so I had to add this back in - without it my SPICE connections gives:

```
Failed to set console: Call to virDomainUpdateDeviceFlags failed: internal error cannot find existing graphics device to modify
```

domcleal replied:
Does this mean it won't work if you have passwords enabled, and are using SPICE?  As far as I can see, we'd configure the client to use a random_password, but skip reconfiguring the VM.

domcleal replied:
Got it, you need to specify `:type` as well so it's able to find the existing graphics device.  I'd suggest using `vm.display[:type]` so we don't try to change the existing type.

GregSutcliffe commented:
Fixed test failures (I hope)

domcleal commented on app/views/compute_resources/show.html.erb:
```diff
@@ -32,6 +32,8 @@
         <td><%= @compute_resource.url %></td>
       </tr>
+      <% # optional extra display elements based on type %>
+      <%= render "compute_resources/show/#{@compute_resource.provider.downcase}" %>
```
[done] You'll need to add empty .erb templates for non-libvirt providers, or a rescue.

domcleal commented on app/views/compute_resources/show/_libvirt.html.erb:
```diff
@@ -0,0 +1,8 @@
+<tr>
+  <td><%= _("Display Type") %></td>
```
Nitpick: this and "Display Passwords" should probably have the second word in lower case.

domcleal commented on app/views/compute_resources/show/_libvirt.html.erb:
```diff
@@ -0,0 +1,8 @@
+  <td><%= @compute_resource.attrs[:display] %></td>
+</tr>
+<tr>
+  <td><%= _("Display Passwords") %></td>
```
Display?  Isn't this the setting for random passwords or not?

domcleal replied:
Ah got it.  Maybe "Set display password" or "Set console password"?  This needs to be consistent with the field on the edit page.

domcleal commented on app/models/compute_resource.rb:
```diff
@@ -190,7 +190,13 @@ def sanitize_url
   end
 
+  def self.set_password?
+    # Overridden for some models
```
Better to move the attrs implementation of set_password here I think.

domcleal commented on app/models/compute_resource.rb:
```diff
@@ -190,7 +190,13 @@ def sanitize_url
     self.url.chomp!("/") unless url.empty?
   end
 
+  def self.set_password?
```
`set_console_password` would be more descriptive, as it's not a password for the VM itself

domcleal commented on app/views/compute_resources/form/_libvirt.html.erb:
```diff
@@ -1,5 +1,9 @@
 <%= text_f f, :url, :class => "input-xlarge", :help_block => _("e.g. qemu://host.example.com/system") %>
 
+<%= select_f f,   :display_type, %w[VNC SPICE],:upcase, :to_s, { }, :label => _("Display Type") %>
```
Lower case 'T'

domcleal commented on app/models/compute_resources/foreman/model/libvirt.rb:
```diff
@@ -3,6 +3,23 @@ class Libvirt < ComputeResource
+    end
+
+    def set_password?
+      self.attrs[:setpw] == "1"
```
The default has changed now so passwords aren't set, but I don't think we should do that.  I think this should return true if `attrs[:setpw]` isn't set.

domcleal commented on app/models/compute_resources/foreman/model/libvirt.rb:
```diff
@@ -3,6 +3,23 @@ class Libvirt < ComputeResource
+    end
+
+    def set_password=(setpw)
+      self.attrs[:setpw] = setpw
```
Squash values here into booleans, so we're not doing string comparisons in the getter.

GregSutcliffe commented:
@domcleal i think that should do it - adding empty templates felt cleaner, but happy to move to a rescue if need be

domcleal commented on app/views/compute_resources/show/_libvirt.html.erb:
```diff
@@ -0,0 +1,8 @@
+<tr>
+  <td><%= _("Display type") %></td>
+  <td><%= @compute_resource.attrs[:display] %></td>
```
This should use the display_type getter.

domcleal commented on app/models/compute_resources/foreman/model/libvirt.rb:
```diff
@@ -134,12 +143,19 @@ def disconnect
     end
 
     def vm_instance_defaults
+      display = self.attrs[:display].present? ? self.attrs[:display].downcase : 'vnc'
```
This should be in and it should use the display_type getter, so the default is visible in the UI

GregSutcliffe replied:
"This should be in" what? :)

domcleal replied:
Yes, poor English was.

This logic should be in the display_type getter, and this code should use the getter.

GregSutcliffe replied:
I'm guessing you mean the logic should be in the getter and the getter alone used in the defaults hash. I _almost_ agree - I'm going to keep the downcase in the defaults, since it's only relevant for sending the right string to Fog.

domcleal replied:
Agreed.

domcleal commented on app/models/compute_resource.rb:
```diff
@@ -180,6 +180,14 @@ def available_images
+  end
+
+  def set_console_password=(setpw)
+    self.attrs[:setpw] = setpw.to_i
```
[done] If you're assuming this is an int, the model should have a validator

GregSutcliffe replied:
Done as per later comments, validators on serialized things are hard ;)

domcleal commented on app/views/compute_resources/form/_libvirt.html.erb:
```diff
@@ -1,5 +1,10 @@
 <%= text_f f, :url, :class => "input-xlarge", :help_block => _("e.g. qemu://host.example.com/system") %>
 
+<%= select_f f,   :display_type, %w[VNC SPICE],:upcase, :to_s, { }, :label => _("Display type") %>
+<%= checkbox_f f, :set_console_password, :checked => f.object.set_console_password?,
```
[done] I think this also needs to be in the VMware view, as it uses random_password too.

domcleal commented:
Thanks, this is looking good, with some useful infrastructure that we can share across CRs.  Aside from the minor notes above, would you be able to add tests for vm_instance_defaults and/or random_password for the new behaviour?  Testing a CR should be trivial as we already have fixtures in place for libvirt.

GregSutcliffe commented:
@domcleal done - I think it's ok to use `send` to access protected methods in tests.

domcleal commented on app/models/compute_resource.rb:
```diff
@@ -180,6 +180,19 @@ def available_images
+    if num=setpw.to_i
+      self.attrs[:setpw] = num
+    end
+  rescue
```
`.to_i` doesn't raise any exceptions, so I think just revert this to `self.attrs[:setpw] = num.to_i` and ignore my validator suggestion, sorry for the noise.

domcleal commented on app/views/compute_resources/show/_libvirt.html.erb:
```diff
@@ -0,0 +1,8 @@
+</tr>
+<tr>
+  <td><%= _("Console passwords") %></td>
+  <td><%= @compute_resource.set_console_password? ? _("True") : _("False") %></td>
```
Nitpick, "Enabled"/"Disabled" might be slightly more user-friendly.  I was trying to find another example in Foreman for consistency, but this was the best I could come up with.

domcleal commented:
I hit an error here because app/helpers/compute_resources_vms_helper.rb in spice_data_attributes adds :ca_cert into the options, but this is only populated for oVirt, not libvirt, so `URL.escape(nil)` breaks.

GregSutcliffe commented:
@domcleal I can't reproduce that, Spice + Console Password is working fine for me here in Chrome now. I think adding `if console[:ca_cert].present?` is probably ok in that helper though, if you want to try it.

domcleal commented:
Confirmed, your recent change fixed the problem in Firefox.

Merged as b760d48dd with a couple of fixups.  Thanks for the patch @GregSutcliffe!
